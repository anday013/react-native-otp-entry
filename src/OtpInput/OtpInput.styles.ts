import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  container: {
    width: "100%",
    flexDirection: "row",
    justifyContent: "space-between",
  },
  codeContainer: {
    borderWidth: 1,
    borderRadius: 12,
    borderColor: "#DFDFDE",
    height: 60,
    width: 44,
    justifyContent: "center",
    alignItems: "center",
  },
  codeText: {
    fontSize: 28,
  },
  hiddenInput: {
    ...StyleSheet.absoluteFillObject,
    opacity: 0,
  },
  stick: {
    width: 2,
    height: 30,
    backgroundColor: "green",
  },
});
