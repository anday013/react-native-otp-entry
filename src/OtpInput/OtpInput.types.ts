import { ColorValue, TextInputProps, TextProps, TextStyle, ViewStyle } from "react-native";

export interface OtpInputProps {
  numberOfDigits?: number;
  autoFocus?: boolean;
  focusColor?: ColorValue;
  onTextChange?: (text: string) => void;
  onFilled?: (text: string) => void;
  onFocus?: () => void;
  onBlur?: () => void;
  blurOnFilled?: boolean;
  hideStick?: boolean;
  focusStickBlinkingDuration?: number;
  secureTextEntry?: boolean;
  theme?: Theme;
  disabled?: boolean;
  textInputProps?: TextInputProps;
  textProps?: TextProps;
  type?: "alpha" | "numeric" | "alphanumeric";
  placeholder?: string;
}

export interface OtpInputRef {
  clear: () => void;
  focus: () => void;
  setValue: (value: string) => void;
}

export interface Theme {
  containerStyle?: ViewStyle;
  /**
   * @deprecated Use `containerStyle` instead
   */
  inputsContainerStyle?: ViewStyle;
  pinCodeContainerStyle?: ViewStyle;
  filledPinCodeContainerStyle?: ViewStyle;
  pinCodeTextStyle?: TextStyle;
  focusStickStyle?: ViewStyle;
  focusedPinCodeContainerStyle?: ViewStyle;
  disabledPinCodeContainerStyle?: ViewStyle;
  placeholderTextStyle?: TextStyle;
}
