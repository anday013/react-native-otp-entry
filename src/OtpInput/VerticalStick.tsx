import * as React from "react";
import { memo, useEffect, useRef } from "react";
import { Animated, ColorValue, View, ViewStyle } from "react-native";
import { styles } from "./OtpInput.styles";

interface VerticalStickProps {
  focusColor?: ColorValue;
  style?: ViewStyle;
  focusStickBlinkingDuration?: number;
}

export const VerticalStick: React.FC<VerticalStickProps> = memo(
  ({ focusColor, style, focusStickBlinkingDuration = 350 }) => {
    const opacityAnim = useRef(new Animated.Value(1)).current;

    useEffect(() => {
      Animated.loop(
        Animated.sequence([
          Animated.timing(opacityAnim, {
            toValue: 0,
            useNativeDriver: true,
            duration: focusStickBlinkingDuration,
          }),
          Animated.timing(opacityAnim, {
            toValue: 1,
            useNativeDriver: true,
            duration: focusStickBlinkingDuration,
          }),
        ]),
        {
          iterations: -1,
        }
      ).start();
    }, []);

    return (
      <Animated.View style={{ opacity: opacityAnim }}>
        <View
          style={[styles.stick, focusColor ? { backgroundColor: focusColor } : {}, style]}
          testID="otp-input-stick"
        />
      </Animated.View>
    );
  }
);
