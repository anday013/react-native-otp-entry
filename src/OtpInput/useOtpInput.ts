import { useRef, useState } from "react";
import { Keyboard, TextInput } from "react-native";
import { OtpInputProps } from "./OtpInput.types";

export const useOtpInput = ({
  onTextChange,
  onFilled,
  numberOfDigits = 6,
  disabled,
  autoFocus = true,
}: OtpInputProps) => {
  const [text, setText] = useState("");
  const [hasCursor, setHasCursor] = useState(autoFocus);
  const inputRef = useRef<TextInput>(null);
  const focusedInputIndex = text.length;

  const handlePress = () => {
    // To fix bug when keyboard is not popping up after being dismissed
    if (!Keyboard.isVisible()) {
      Keyboard.dismiss();
    }
    inputRef.current?.focus();
  };

  const handleTextChange = (value: string) => {
    if (disabled) return;
    setText(value);
    onTextChange?.(value);
    if (value.length === numberOfDigits) {
      onFilled?.(value);
    }
  };

  const setTextWithRef = (value: string) => {
    const normalizedValue = value.length > numberOfDigits ? value.slice(0, numberOfDigits) : value;
    handleTextChange(normalizedValue);
  };

  const clear = () => {
    setText("");
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const handleFocus = () => {
    setHasCursor(true);
  };

  const handleBlur = () => {
    setHasCursor(false);
  };

  return {
    models: { text, inputRef, focusedInputIndex, hasCursor },
    actions: { handlePress, handleTextChange, clear, focus, handleFocus, handleBlur },
    forms: { setText, setTextWithRef },
  };
};
