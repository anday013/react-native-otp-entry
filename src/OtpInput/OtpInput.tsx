import { forwardRef, useImperativeHandle } from "react";
import { Platform, Pressable, Text, TextInput, View } from "react-native";
import { styles } from "./OtpInput.styles";
import { OtpInputProps, OtpInputRef } from "./OtpInput.types";
import { VerticalStick } from "./VerticalStick";
import { useOtpInput } from "./useOtpInput";

export const OtpInput = forwardRef<OtpInputRef, OtpInputProps>((props, ref) => {
  const {
    models: { text, inputRef, focusedInputIndex, hasCursor },
    actions: { clear, handlePress, handleTextChange, focus, handleFocus, handleBlur },
    forms: { setTextWithRef },
  } = useOtpInput(props);
  const {
    disabled,
    numberOfDigits = 6,
    autoFocus = true,
    hideStick,
    focusColor = "#A4D0A4",
    focusStickBlinkingDuration,
    secureTextEntry = false,
    theme = {},
    textInputProps,
  } = props;
  const {
    containerStyle,
    inputsContainerStyle,
    pinCodeContainerStyle,
    pinCodeTextStyle,
    focusStickStyle,
    focusedPinCodeContainerStyle,
    filledPinCodeContainerStyle,
    disabledPinCodeContainerStyle,
  } = theme;

  useImperativeHandle(ref, () => ({ clear, focus, setValue: setTextWithRef }));

  const generatePinCodeContainerStyle = (isFocusedInput: boolean, char: string) => {
    const stylesArray = [styles.codeContainer, pinCodeContainerStyle];
    if (focusColor && isFocusedInput) {
      stylesArray.push({ borderColor: focusColor });
    }

    if (focusedPinCodeContainerStyle && isFocusedInput) {
      stylesArray.push(focusedPinCodeContainerStyle);
    }

    if (filledPinCodeContainerStyle && Boolean(char)) {
      stylesArray.push(filledPinCodeContainerStyle);
    }

    if (disabledPinCodeContainerStyle && disabled) {
      stylesArray.push(disabledPinCodeContainerStyle);
    }

    return stylesArray;
  };

  return (
    <View style={[styles.container, containerStyle]}>
      <View style={[styles.inputsContainer, inputsContainerStyle]}>
        {Array(numberOfDigits)
          .fill(0)
          .map((_, index) => {
            const char = text[index];
            const isFocusedInput = index === focusedInputIndex && !disabled && Boolean(hasCursor);

            return (
              <Pressable
                key={`${char}-${index}`}
                disabled={disabled}
                onPress={handlePress}
                style={generatePinCodeContainerStyle(isFocusedInput, char)}
                testID="otp-input"
              >
                {isFocusedInput && !hideStick ? (
                  <VerticalStick
                    focusColor={focusColor}
                    style={focusStickStyle}
                    focusStickBlinkingDuration={focusStickBlinkingDuration}
                  />
                ) : (
                  <Text style={[styles.codeText, pinCodeTextStyle]}>
                    {char && secureTextEntry ? "â€¢" : char}
                  </Text>
                )}
              </Pressable>
            );
          })}
      </View>
      <TextInput
        value={text}
        onChangeText={handleTextChange}
        maxLength={numberOfDigits}
        textContentType="oneTimeCode"
        ref={inputRef}
        autoFocus={autoFocus}
        secureTextEntry={secureTextEntry}
        autoComplete={Platform.OS === "android" ? "sms-otp" : "one-time-code"}
        aria-disabled={disabled}
        editable={!disabled}
        testID="otp-input-hidden"
        onFocus={handleFocus}
        onBlur={handleBlur}
        inputMode={textInputProps?.inputMode || "numeric"}
        keyboardType={textInputProps?.keyboardType || "numeric"}
        {...textInputProps}
        style={[styles.hiddenInput, textInputProps?.style]}
      />
    </View>
  );
});
