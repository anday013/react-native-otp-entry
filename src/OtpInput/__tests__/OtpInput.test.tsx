import { act, fireEvent, render, screen } from "@testing-library/react-native";
import * as React from "react";
import { OtpInput } from "../OtpInput";
import { OtpInputProps, OtpInputRef } from "../OtpInput.types";

const renderOtpInput = (props?: Partial<OtpInputProps>) => render(<OtpInput {...props} />);

describe("OtpInput", () => {
  describe("UI", () => {
    test("should render correctly", () => {
      const tree = renderOtpInput().toJSON();

      expect(tree).toMatchSnapshot();
    });

    test("should render stick if hideStick is false", () => {
      renderOtpInput({ hideStick: false });

      const stick = screen.getByTestId("otp-input-stick");

      expect(stick).toBeTruthy();
    });

    test('should not show values if "secureTextEntry" is true', () => {
      renderOtpInput({ secureTextEntry: true });

      const input = screen.getByTestId("otp-input-hidden");
      fireEvent.changeText(input, "123456");
      const inputs = screen.getAllByTestId("otp-input");

      inputs.forEach((input) => {
        expect(input).toHaveTextContent("â€¢");
      });
    });

    test("should have 6 inputs by default", () => {
      renderOtpInput();

      const inputs = screen.getAllByTestId("otp-input");

      expect(inputs).toHaveLength(6);
    });

    test("should autoFocused by default", () => {
      renderOtpInput();

      const input = screen.getByTestId("otp-input-hidden");

      expect(input.props.autoFocus).toBe(true);
    });

    test('should not focus if "autoFocus" is false', () => {
      renderOtpInput({ autoFocus: false });

      const input = screen.getByTestId("otp-input-hidden");

      expect(input.props.autoFocus).toBe(false);
    });

    test("it should not allow input if disabled is true", () => {
      renderOtpInput({ disabled: true });

      const input = screen.getByTestId("otp-input-hidden");
      fireEvent.changeText(input, "123456");

      const inputs = screen.getAllByTestId("otp-input");
      expect(inputs[0]).not.toHaveTextContent("1");
      inputs.forEach((i) => expect(i).toBeDisabled());

      const hiddenInput = screen.getByTestId("otp-input-hidden");
      expect(hiddenInput).toBeDisabled();
    });

    test("focusColor should not be overridden by theme", () => {
      renderOtpInput({
        focusColor: "#000",
        theme: { pinCodeContainerStyle: { borderColor: "#fff" } },
      });

      const inputs = screen.getAllByTestId("otp-input");

      expect(inputs[0]).toHaveStyle({ borderColor: "#000" });
    });

    test("focusedPinCodeContainerStyle should not be overridden by theme", () => {
      renderOtpInput({
        focusColor: "#000",
        theme: { focusedPinCodeContainerStyle: { borderColor: "#fff", backgroundColor: "green" } },
      });
      const inputs = screen.getAllByTestId("otp-input");
      expect(inputs[0]).toHaveStyle({ borderColor: "#fff", backgroundColor: "green" });
    });

    // Test if the number of rendered inputs is equal to the number of digits
    test.each([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])(
      "should render the correct number of inputs: %i",
      (numberOfDigits) => {
        renderOtpInput({ numberOfDigits: numberOfDigits });

        const inputs = screen.getAllByTestId("otp-input");

        expect(inputs).toHaveLength(numberOfDigits);
      }
    );

    test("filledPinCodeContainerStyle should allow for new style when digit is present", () => {
      renderOtpInput({
        theme: { filledPinCodeContainerStyle: { borderBottomColor: "red" } },
      });

      const input = screen.getByTestId("otp-input-hidden");
      fireEvent.changeText(input, "12");

      const inputs = screen.getAllByTestId("otp-input");
      expect(inputs[0]).toHaveStyle({ borderBottomColor: "red" });
      expect(inputs[1]).toHaveStyle({ borderBottomColor: "red" });
      expect(inputs[2]).not.toHaveStyle({ borderBottomColor: "red" });
    });

    test("disabledPinCodeContainerStyle should allow for new style when input is disabled", () => {
      renderOtpInput({
        disabled: true,
        theme: { disabledPinCodeContainerStyle: { borderBottomColor: "red" } },
      });

      const inputs = screen.getAllByTestId("otp-input");
      expect(inputs[0]).toHaveStyle({ borderBottomColor: "red" });
    });
  });

  describe("Logic", () => {
    test("should split text on screen from the text written in the hidden input", () => {
      const otp = "123456";
      renderOtpInput();

      const input = screen.getByTestId("otp-input-hidden");
      fireEvent.changeText(input, otp);

      act(() => {
        expect(screen.getByText("1")).toBeTruthy();
        expect(screen.getByText("2")).toBeTruthy();
        expect(screen.getByText("3")).toBeTruthy();
        expect(screen.getByText("4")).toBeTruthy();
        expect(screen.getByText("5")).toBeTruthy();
        expect(screen.getByText("6")).toBeTruthy();
        expect(screen.queryByText(otp)).not.toBeTruthy();
      });
    });

    test("ref clear() should clear input", () => {
      const ref = React.createRef<OtpInputRef>();

      render(<OtpInput ref={ref} numberOfDigits={6} />);
      const otp = "1";

      const input = screen.getByTestId("otp-input-hidden");
      fireEvent.changeText(input, otp);

      act(() => {
        ref.current?.clear();
      });
      expect(screen.queryByText("1")).toBeFalsy();
    });

    test("ref setValue() should set input value", () => {
      const ref = React.createRef<OtpInputRef>();

      render(<OtpInput ref={ref} numberOfDigits={6} />);
      const otp = "1";

      act(() => {
        ref.current?.setValue(otp);
      });
      expect(screen.getByText("1")).toBeTruthy();
    });

    test('ref setValue() should set only the first "numberOfDigits" characters', () => {
      const ref = React.createRef<OtpInputRef>();

      render(<OtpInput ref={ref} numberOfDigits={4} />);
      const otp = "123456";

      act(() => {
        ref.current?.setValue(otp);
      });
      expect(screen.getByText("1")).toBeTruthy();
      expect(screen.getByText("2")).toBeTruthy();
      expect(screen.getByText("3")).toBeTruthy();
      expect(screen.getByText("4")).toBeTruthy();
      expect(screen.queryByText("5")).toBeFalsy();
      expect(screen.queryByText("6")).toBeFalsy();
    });
  });
});
