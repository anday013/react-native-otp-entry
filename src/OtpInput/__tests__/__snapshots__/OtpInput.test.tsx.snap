// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OtpInput UI should render correctly 1`] = `
<View
  style={
    [
      {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        undefined,
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderColor": "#DFDFDE",
            "borderRadius": 12,
            "borderWidth": 1,
            "height": 60,
            "justifyContent": "center",
            "width": 44,
          },
          undefined,
          {
            "borderColor": "#A4D0A4",
          },
        ]
      }
      testID="otp-input"
    >
      <View
        collapsable={false}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            [
              {
                "backgroundColor": "green",
                "height": 30,
                "width": 2,
              },
              {
                "backgroundColor": "#A4D0A4",
              },
              undefined,
            ]
          }
          testID="otp-input-stick"
        />
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderColor": "#DFDFDE",
            "borderRadius": 12,
            "borderWidth": 1,
            "height": 60,
            "justifyContent": "center",
            "width": 44,
          },
          undefined,
        ]
      }
      testID="otp-input"
    >
      <Text
        style={
          [
            {
              "fontSize": 28,
            },
            undefined,
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderColor": "#DFDFDE",
            "borderRadius": 12,
            "borderWidth": 1,
            "height": 60,
            "justifyContent": "center",
            "width": 44,
          },
          undefined,
        ]
      }
      testID="otp-input"
    >
      <Text
        style={
          [
            {
              "fontSize": 28,
            },
            undefined,
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderColor": "#DFDFDE",
            "borderRadius": 12,
            "borderWidth": 1,
            "height": 60,
            "justifyContent": "center",
            "width": 44,
          },
          undefined,
        ]
      }
      testID="otp-input"
    >
      <Text
        style={
          [
            {
              "fontSize": 28,
            },
            undefined,
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderColor": "#DFDFDE",
            "borderRadius": 12,
            "borderWidth": 1,
            "height": 60,
            "justifyContent": "center",
            "width": 44,
          },
          undefined,
        ]
      }
      testID="otp-input"
    >
      <Text
        style={
          [
            {
              "fontSize": 28,
            },
            undefined,
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderColor": "#DFDFDE",
            "borderRadius": 12,
            "borderWidth": 1,
            "height": 60,
            "justifyContent": "center",
            "width": 44,
          },
          undefined,
        ]
      }
      testID="otp-input"
    >
      <Text
        style={
          [
            {
              "fontSize": 28,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
  <TextInput
    autoComplete="one-time-code"
    autoFocus={true}
    editable={true}
    inputMode="numeric"
    keyboardType="numeric"
    maxLength={6}
    onChangeText={[Function]}
    secureTextEntry={false}
    style={
      {
        "bottom": 0,
        "left": 0,
        "opacity": 0.01,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="otp-input-hidden"
    textContentType="oneTimeCode"
    value=""
  />
</View>
`;
